### PR TITLE
(DOCSP-19488): Flexible Sync Documentation - JS Usage Examples

### PR DESCRIPTION
**Pull request by mongomoe**
### Jira

- https://jira.mongodb.org/browse/DOCSP-19488

### Staged Changes (Requires MongoDB Corp SSO)

- [Usage Examples > Open a Flexible Synced Realm](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-19488-flexible-sync-JS/sdk/node/examples/open-and-close-a-realm/#open-a-flexible-synced-realm) (The only new content in this page is the content under the heading "Open a Flexible Synced Realm")
- [Usage Examples > Flexible Sync](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-19488-flexible-sync-JS/sdk/node/examples/flexible-sync/) (Completely new page, please review the entire page)
- [Usage Examples > Sync Changes Between Devices # Prerequisites](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-19488-flexible-sync-JS/sdk/node/examples/sync-changes-between-devices/#prerequisites)  (The only new content in this page is the content under the heading "Prerequisites", however, the old "Open a Synced Realm" content has been removed from this page in favor of keeping it consistently in the "Open and Close a Realm" page)

### Note to Technical Reviewer:
- If you're just looking to provide a technical review, rather than both a technical and copy/wording review, you can see & comment on just the code easily (and avoid the generated files and documentation files) by looking at the examples/node/Examples/flexible-sync.js file (see: https://github.com/mongodb/docs-realm/pull/1539/files#diff-7cecd1eac97bf9ccab7948ea72f5791b2c3c865bc24e1b9fd164e65f4cf5fdefR1)

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: examples/node/Examples/flexible-sync.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    // :code-block-start: remove-single-subscription
+    subscriptions.update((mutableSubscriptionsInstance) => {
+      // remove a subscription with a specific query
+      mutableSubscriptionsInstance.remove(tasks.filtered('owner == "Ben"'));
+    });
+    // :code-block-end:
+
+    // :code-block-start: remove-subscription-by-name
+    subscriptions.update((mutableSubscriptionsInstance) => {
+      // remove a subscription with a specific query
```
Comment (tomduncalf) [inline examples/node/Examples/flexible-sync.js:89]: This should say `specific name`

File: examples/node/Examples/flexible-sync.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      mutableSubscriptionsInstance.remove("longRunningTasksSubscription");
```
Comment (tomduncalf) [inline examples/node/Examples/flexible-sync.js:90]: This should be `removeByName`

File: source/sdk/node/examples/flexible-sync.txt [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+.. note::
+ 
+   Attempting to update a subscription that has the
+   ``SubscriptionOptions.throwOnUpdate`` field set to true, throw an exception.
+
+Remove Subscriptions
+~~~~~~~~~~~~~~~~~~~~
+To remove subscriptions from the subscription set, you can:
+
+- Remove a single subscription with the given query
```
Comment (tomduncalf) [inline source/sdk/node/examples/flexible-sync.txt:163]: You can also remove a subscription if you have a reference to that subscription: https://github.com/realm/realm-js/blob/td/flexible-sync-final-squash/types/index.d.ts#L858

e.g.

```
let sub;
subs.update(mutableSubs => {
  sub = mutableSubs.add(realm.objects("Person"));
});

// later..
subs.removeSubscription(sub);
```

but if you chose to leave this out as its an advanced use case, that's fine!

Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt:179]: Added this as an example under the subheading "Remove a Subscription by Reference"

File: examples/node/Examples/flexible-sync.js (hunk 2 above)
Comment (mongomoe) [reply examples/node/Examples/flexible-sync.js]: Fixed: replaced ``remove()`` with ``removeByName()`` here 👍

File: examples/node/Examples/flexible-sync.js (hunk 1 above)
Comment (mongomoe) [reply examples/node/Examples/flexible-sync.js]: Good catch, fixed now 👍

File: source/sdk/node/examples/flexible-sync.txt [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+but a different query throws an exception.
+
+You can add a single query, or multiple queries within a ``subscription.update``
+transaction. In the example below, we subscribe to the queries we created
+earlier. 
+
+.. literalinclude:: /examples/generated/node/flexible-sync.codeblock.subscribe-to-queryable-fields.js
+    :language: javascript
+
+Check the Status of Subscriptions
```
[the remark this answers is not in the document]
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt:121]: Added the "waitForSynchronization" method now 👍 Let me know if I worded the explanation correctly.

File: source/sdk/node/examples/flexible-sync.txt [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+You can use this snapshot to add queries to this list of subscriptions and
+update existing subscriptions, as shown in the examples below.
+
+Add a Query to the List Of Subscriptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
+Within a Flexible Sync ``transaction`` you can add queries to your list of subscriptions.
+
+Queries must be added to the list of existing subscriptions within a
+``transaction``. Create a transaction by passing a callback function ``update()`` on
```
[the remark this answers is not in the document]
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt]: That's a good suggestion, I've replaced the existing wording with your suggestion, but rephrased it a bit & made it a list to make it even more easy to read: 👍 

```
You can add queries to your subscriptions list. To do so, perform the following: 

1. Create a transaction by passing a callback function to the ``update()`` method of your subscriptions snapshot. The callback function provides a mutable instance of the subscriptions as an argument. 
2. Within the callback function, call the ``add()`` method on the mutable subscription instance to add a query to the subscription.
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+syncs to the client device.
+
+.. include:: /includes/note-flexible-sync-preview.rst
+
+Subscribe to Queryable Fields
+-----------------------------
+
+When configuring Flexible Sync on the backend, you specify which fields
+your client application can query. In the client application, use the 
+``subscriptions`` API to manage a set of subscriptions to specific queries on
```
[the remark this answers is not in the document]
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt]: Fixed: changed to ``Subscriptions`` 👍

**Review by tomduncalf — APPROVED** [review]
👍

File: source/sdk/node/examples/flexible-sync.txt [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+.. example::
+
+  In the following example, several queries are created to subscribe to:
+
+  - A filtered list of completed tasks that have taken over 120 ``progressMinutes``
+  - A filtered list of completed tasks
+
+  Notice that queries that you subscribe to using Flexible Sync are
+  syntactically the same as :ref:`querying a {+realm+} <node-read-operations>`
+  and that the queries below will create an overlapping result set.
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:80]: Will the example be updated? Right now it doesn't look like we're using flexible sync, just filtering objects.

File: source/sdk/node/examples/flexible-sync.txt [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+Get Subscriptions
+~~~~~~~~~~~~~~~~~
+To get a list of read-only snapshot of subscriptions, call
+``getSubscriptions()`` method on the {+realm+}. 
+
+.. literalinclude:: /examples/generated/node/flexible-sync.codeblock.get-subscriptions.js
+    :language: javascript
+
+You can use this snapshot to add queries to this list of subscriptions and
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:93]: Which list?
```suggestion
You can use this snapshot to add queries to the list of subscriptions and
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+  Notice that queries that you subscribe to using Flexible Sync are
+  syntactically the same as :ref:`querying a {+realm+} <node-read-operations>`
+  and that the queries below will create an overlapping result set.
+
+  .. literalinclude:: /examples/generated/node/flexible-sync.codeblock.create-queries-to-subscribe-to.js
+     :language: javascript
+
+Get Subscriptions
+~~~~~~~~~~~~~~~~~
+To get a list of read-only snapshot of subscriptions, call
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:87]: "list of read-only snapshots of subscriptions" is confusing. Especially given that we say we can use the snapshot to add queries, so I don't understand what's "read-only" about it. Could you clarify?

File: source/sdk/node/examples/flexible-sync.txt [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+update existing subscriptions, as shown in the examples below.
+
+Add a Query to the List Of Subscriptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
+You can add queries to your subscriptions list. To do so, perform the following: 
+
+#. Create a transaction by passing a callback function to the ``update()`` method of your subscriptions snapshot. The callback function provides a mutable instance of the subscriptions as an argument. 
+#. Within the callback function, call the ``add()`` method on the mutable subscription instance to add a query to the subscription.
+
+The ``add()`` method on the mutable subscription instance takes a ``query``, and a
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:103]:
```suggestion
The ``add()`` method on the mutable subscription instance takes a ``query`` and a
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+subscription options object. The subscription options include:
+
+- a ``name`` string field
+- a ``throwOnUpdate`` boolean field.
+
+If ``throwOnUpdate`` is false or not defined, adding a subscription with an
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:109]: May as well use the JS terminology
```suggestion
If ``throwOnUpdate`` is false or undefined, adding a subscription with an
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+existing name will replace the existing query with the new query. However, if
+you set ``throwOnUpdate`` to true, adding a subscription with an existing name
+but a different query throws an exception.
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:112]: Minor style point:

Currently, the two sentences are like this:
- First sentence: If X is false, ...
- Second sentence: However, if **you** set...

But you could parallelize this better:
- If X is false, ...
- However, if X is true, ...

(And I would actually remove the "However", since the parallel makes it pretty clear it's the opposite!)

File: source/sdk/node/examples/flexible-sync.txt [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+.. literalinclude:: /examples/generated/node/flexible-sync.codeblock.subscribe-to-queryable-fields.js
+    :language: javascript
+
+Check the Status of Subscriptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+You can use the ``waitForSynchronization()`` method to wait for the server to
+acknowledge this set of subscriptions and return the matching objects: 
+
+The ``waitForSynchronization()`` method returns a promise which is resolved when
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:126]: That instead of which (specific instance instead of general), and also could go active on that:
```suggestion
The ``waitForSynchronization()`` method returns a promise that resolves when
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+synchronization is complete, or is rejected if there is an error during
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:127]: With previous suggestion, plus remove comma since subject did not change:
```suggestion
synchronization is complete or rejects if there is an error during
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+synchronisation.
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:128]: American spelling:
```suggestion
synchronization.
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+   The ``waitForSynchronization()`` method throws if there is an error, so we
+   recommend wrapping it in a :mdn:`try...catch
+   <Web/JavaScript/Reference/Statements/try...catch>` statement.
+
+.. literalinclude:: /examples/generated/node/flexible-sync.codeblock.wait-for-synchronization.js
+    :language: javascript
+
+In addition to the ``waitForSynchronization()`` method, you can check the
+subscription state to see if the server has acknowledged the subscription and
+the device has downloaded the data has locally.
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:141]: Stray "has"?
```suggestion
the device has downloaded the data locally.
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+
+You can use subscription state to:
+
+- Trigger error handling
+- Show if the transaction is pending or has completed
+- Find out when a subscription set is superseded, and you should obtain a
+  new instance of the subscription set to write a subscription change
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:148]: I think I commented similarly on the other SDK PR: I find this surprising. The first two sound totally normal and the third one throws a lot of new terms at you. The structure of the sentence is hard to parse. Maybe instead of "and you should", you could say "which means you should"?

File: source/sdk/node/examples/flexible-sync.txt [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+Update Subscriptions with a New Query
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+You can update a named subscription with a new query. To update a subscriptions
+query, pass the new query and a subscription option with the name of the
+subscription that you want to update to the
+``mutableSubscriptionsInstance.add()`` method. Like adding a new subscription,
+you must update a subscription within a transaction by calling
+``subscriptions.update()`` method.
+
+In the following example, long running tasks are re-defined to be any tasks that
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:164]:
```suggestion
In the following example, long-running tasks are re-defined to be any tasks that
```

File: source/sdk/node/examples/flexible-sync.txt [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+do so, execute a transaction on the subscriptions set. Within the transaction,
+pass the reference variable to the ``removeSubscription()`` method on the
+mutable subscriptions instance.
+
+.. literalinclude:: /examples/generated/node/flexible-sync.codeblock.remove-subscription-by-reference.js
+    :language: javascript
+
+Remove All Subscriptions of an Object Type
+``````````````````````````````````````````
+To remove all subscriptions of a specific object type, execute a transaction on
```
Comment (cbush) [inline source/sdk/node/examples/flexible-sync.txt:221]: Should we say "subscriptions OF a type" or do we mean "subscriptions ON a type"?

**Review by cbush — APPROVED** [review]
Looks great, but I have lots of comments I would like to see addressed before you merge.

File: source/sdk/node/examples/flexible-sync.txt [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+A Realm subscription in JavaScript has:
+
+- A created and updated date 
+- A name
+- An object type
+- A query string 
+
+You can specify a string name for your subscription. If you do not give your
+subscription a name, the name returns :mdn:`null
+<Web/JavaScript/Reference/Global_Objects/null>`.
```
[the remark this answers is not in the document]
Comment (tomduncalf) [reply source/sdk/node/examples/flexible-sync.txt:64]: This behaviour was changed, as far as I know all SDKs return `null` instead of query string now as we decided it didn't make that much sense to return the query string for the name

File: source/sdk/node/examples/flexible-sync.txt (hunk 9 above)
Comment (tomduncalf) [reply source/sdk/node/examples/flexible-sync.txt:87]: Agree the wording seems off – just to clarify from a technical point of view, the return value from `getSubscriptions()` is a read-only, immutable snapshot of the current subscriptions at that point in time. 

In order to modify the set of subscriptions, you need to call `subs.update(mutableSubs => { mutableSubs.add... })` – so you can only modify the mutable instance you get in the callback, the `Subscriptions` instance returned from `getSubscriptions` does not have `add`/`remove` etc methods.

This is by design to try to guide users towards updating their subscriptions in a batch, because each update call is quite a heavy process on the server.

File: source/sdk/node/examples/flexible-sync.txt (hunk 17 above)
Comment (tomduncalf) [reply source/sdk/node/examples/flexible-sync.txt:148]: Yeah, the superseded case is quite hard to explain and I guess sort of an edge case for users. I'll try to get a good explanation of it together for you if that's helpful?

File: source/sdk/node/examples/flexible-sync.txt (hunk 20 above)
[the remark this answers is not in the document]
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt:74]: changed to "is set to null".

File: source/sdk/node/examples/flexible-sync.txt [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+
+When you create a subscription, Realm looks for data matching a query on a
+specific object type. In your Flexible Sync subscriptions, you can have
+subscriptions on several different object types or several queries on the same
+object type.
+
+.. example::
+
+  In the following example, several queries are created to subscribe to:
```
[the remark this answers is not in the document]
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt:83]: 👍  took out the word "several"

File: source/sdk/node/examples/flexible-sync.txt (hunk 7 above)
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt:90]: This example is supposed to just be storing a reference to the query, in the example below we actually subscribe to the queries using the reference.

File: source/sdk/node/examples/flexible-sync.txt (hunk 8 above)
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt]: fixed locally 👍

File: source/sdk/node/examples/flexible-sync.txt (hunk 9 above)
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt:97]: Altered the text to be "To get a snapshot of current subscriptions, call ``getSubscriptions()`` method
on the {+realm+}."

File: source/sdk/node/examples/flexible-sync.txt (hunk 19 above)
Comment (mongomoe) [reply source/sdk/node/examples/flexible-sync.txt:231]: Replaced "of" to "on" 👍
